vru_use_case_scripts/process_duplicates.py

```diff
@@ -0,0 +1,109 @@
+import rosbag
+import csv
+import pandas as pd
+import numpy as np
+import matplotlib.pyplot as plt
+import argparse
+from pathlib import Path
+from enum import Enum
+
+
+class ObjectType(Enum):
+    UNKNOWN = 0
+    SMALL_VEHICLE = 1
+    LARGE_VEHICLE = 2
+    MOTORCYCLE = 3
+    PEDESTRIAN = 4
+    def __str__(self):
+        return self.name
+
+
+def plot_duplicates(input_file):
+    # plot duplicate durations of each Object over time
+    df = pd.read_csv(input_file)
+    figures_dir = Path("duplicates_plots")
+    figures_dir.mkdir(exist_ok=True)
+    grouped = df.groupby('ObjID')
+
+    # Plot duration over timestamp for each ObjID
+    for obj_id, group in grouped:
+        obj_type = group['ObjType'].iloc[0]  # Get the ObjType for the current ObjIDMain
+        fig, ax = plt.subplots()
+        ax.stem(group['Timestamp(ms)'], group['Duration'], basefmt=" ", markerfmt=' ', use_line_collection=True)
+        ax.set_xlim(left=0)  # Set x-axis to start at zero
+        ax.set_ylim(bottom=0)  # Set y-axis to start at zero
+        plt.axhline(y=1000, color='r', linestyle='--', label='Cutoff')
+        plt.xlabel('Timestamp(ms)')
+        plt.ylabel('Duration')
+        plt.title(f'Duration over Timestamp for ObjID: {obj_id} with type: {obj_type}')
+        plt.legend()
+        # plt.show()
+        plt.savefig(figures_dir / f"duplicates_{obj_id}.png")
+
+def get_duplicate_duration(input_file, output_file):
+    # Caluclate the duration each duplicate lasts
+    df = pd.read_csv(input_file)
+    df.sort_values(by=['ObjID', 'Timestamp(ms)'], inplace=True)
+    df['TimeDiff'] = df.groupby('ObjID')['Timestamp(ms)'].diff()
+    # Create a mask to identify interruptions in the ObjIDMain sequence
+    mask = (df['TimeDiff'] <= 100)
+    df['Mask'] = mask
+    sum_timediff = 0
+    # calculate the duration each duplicate lasts
+    for index, row in df.iterrows():
+        if row['Mask']:
+            sum_timediff += row['TimeDiff']
+        else:
+            sum_timediff = 0
+        df.at[index, 'Duration'] = sum_timediff
+
+    # Reset the index of the DataFrame
+    df.reset_index(drop=True, inplace=True)
+    df.to_csv(output_file, index=False)
+
+
+def remove_main_obj(input_file, output_file):
+    # The main detections are removed from the detected objects by removing the first instance of each object ID
+    df = pd.read_csv(input_file)
+    # Group by 'Timestamp(ms)' and drop the first row for each group
+    df = df.groupby('Timestamp(ms)').apply(lambda x: x.iloc[1:]).reset_index(drop=True)
```

Comment:
I think I understand what this logic is trying to do here. It is trying to remove first instance of each object at each time step so that if there were 2 same ID objects, it would be 1. And when we parse it later, if a timestamp has 1 or more same id object (although cant be more than 1 due to `drop_duplicates()` step) , then it means there were duplicates. However, the logic seems to not be considering the ID. I ran the script on a sample data. Timestamp 24776 should also remove id `223`

Reply:
Left is all_detected_object.csv, right is duplicates.csv
![duplicates_data_misgh](https://github.com/usdot-fhwa-stol/carma-analytics-fotda/assets/20613282/2d8f5c3e-f44b-4035-b5ee-fa5582627107)

Reply:
What you summarized here is correct. Once the first instance of each detected object is removed, what is left is all duplicates. The drop_duplicate is then run to remove any cases of multiple duplicates at the same msg time.
Thanks for catching that issue, made a change. Please check again